pkg/kubectl/cmd/run.go

```diff
@@ -276,9 +288,35 @@ func Run(f *cmdutil.Factory, cmdIn io.Reader, cmdOut, cmdErr io.Writer, cmd *cob
 				ResourceNames(mapping.Resource, name).
 				Flatten().
 				Do()
-			return ReapResult(r, f, cmdOut, true, true, 0, -1, false, mapper)
+			err = ReapResult(r, f, cmdOut, true, true, 0, -1, false, mapper)
+			if err != nil {
+				return err
+			}
+		}
+
+		if !waitForExitCode {
+			return nil
+		}
+
```

Comment:
Because you have to perform the removal no matter what. If you move this switch block up, it would return too early and the remove wouldn't happen.